### ISSUE TITLE
arm64 "at s1e1r, x0" instruction produces segmentation dump

### ISSUE THREAD

**Issue by wafgo** [avast/retdec#998]
Hi guys,

first of all, thank you for the awesome tool. Great work!

Though I found some issue during decompilation of some aarch64/arm64 System Software (Kernel) which we use at the company. 

I have created a simple program with one simple instruction which produces this problem, so it should be easy for you guys to track the reason for the segdump  correctly. 

I also fixed it on my machine locally, but I am not sure whether it is the correct way of doing it (or maybe rephrase it: I dont understand enough of your software to be apply this patch confidently).

I have build locally the retdec-decompiler from your github revision: **18b434f02b64bfb85232525ca10a6026b5d1f54f**

**Tool Version follows:**
```shell
local@machine:~/devel/github/retdec$ ~/devel/retdec/bin/retdec-decompiler --version
RetDec version :  v4.0-346-g18b434f0
Commit hash    :  18b434f02b64bfb85232525ca10a6026b5d1f54f
Build date     :  2021-08-09T15:02:19Z

```
Here comes my example project

```shell

#print project files
local@machine:~/devel/github/retdec_bug$ ls -la
total 20
drwxrwxr-x  2 local local 4096 Aug 10 15:36 ./
drwxrwxr-x 35 local local 4096 Aug 10 15:12 ../
-rw-rw-r--  1 local local  247 Aug 10 15:35 _at_test.ld
-rw-rw-r--  1 local local   74 Aug 10 15:34 _at_test.S
-rw-rw-r--  1 local local  530 Aug 10 15:25 Makefile

# print assembly
local@machine:~/devel/github/retdec_bug$ cat _at_test.S
.section ".text","ax"

.global _start        
_start:        at s1e1r, x0

# print linker script
local@macine:~/devel/github/retdec_bug$ cat _at_test.ld
OUTPUT_FORMAT("elf64-littleaarch64")
OUTPUT_ARCH(aarch64)
ENTRY(_start)

MEMORY {
    BRAM (rwx): ORIGIN = 0xEB100180, LENGTH = 0x30000
}

SECTIONS
{
    .text : {
        __text_start = .;
        *(.text*)
        __text_end = .;
     } >BRAM
}

# print makefile
local@machine:~/devel/github/retdec_bug$ cat Makefile 
TARGET:=at_test.elf

OBJS:=_at_test.o

all: $(TARGET)

clean:
        rm $(OBJS) $(TARGET)

$(TARGET): $(OBJS)
        /home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-ld -nostartfiles -nodefaultlibs -nostdlib -g -T _at_test.ld -o $@ $?

%.o: %.c
        /home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-gcc -c -march=armv8-a -ggdb3 -Og -mtune=cortex-a57.cortex-a53 -o $@ $<

%.o: %.S
        /home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-gcc -c -march=armv8-a -ggdb3 -Og -mtune=cortex-a57.cortex-a53 -o $@ $<

# build project
local@machine:~/devel/github/retdec_bug$ make
/home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-gcc -c -march=armv8-a -ggdb3 -Og -mtune=cortex-a57.cortex-a53 -o _at_test.o _at_test.S
/home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-ld -nostartfiles -nodefaultlibs -nostdlib -g -T _at_test.ld -o at_test.elf _at_test.o

# dissassemble elf to be sure everything is correctly assembled
local@machine:~/devel/github/retdec_bug$ /home/local/devel/cross/aarch64-gcc-7.4.1/bin/aarch64-linux-gnu-objdump -d at_test.elf

at_test.elf:     file format elf64-littleaarch64


Disassembly of section .text:

00000000eb100180 <__text_start>:
    eb100180:   d5087800        at      s1e1r, x0

# try to decompile using retdec_decompiler
local@BHD6QPUW:~/devel/github-conti/retdec_bug$ ~/devel/retdec/bin/retdec-decompiler at_test.elf
Running phase: Unpacking ( 0.05s )
No matching plugins found for 'gc'
Running phase: Initialization ( 0.05s )
Running phase: Providers initialization ( 0.06s )
Running phase: Input binary to LLVM IR decoding ( 0.33s )
 #0 0x00005584046ffc8a llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/home/local/devel/retdec/bin/retdec-decompiler+0x12f7c8a)
 #1 0x00005584046fd864 llvm::sys::RunSignalHandlers() (/home/local/devel/retdec/bin/retdec-decompiler+0x12f5864)
 #2 0x00005584046fd9a2 SignalHandler(int) (/home/local/devel/retdec/bin/retdec-decompiler+0x12f59a2)
 #3 0x00007f5cfaad3980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #4 0x0000558403a28e0c llvm::Value::getType() const /home/local/devel/github/retdec/build/external/src/llvm-project/include/llvm/IR/Value.h:245:0
 #5 0x00005584040f4f96 retdec::capstone2llvmir::Capstone2LlvmIrTranslatorArm64_impl::extractVectorValue(llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&, cs_arm64_op&, llvm::Value*) /home/local/devel/github/retdec/build/../src/capstone2llvmir/arm64/arm64.cpp:375:0
 #6 0x00005584040f6c08 retdec::capstone2llvmir::Capstone2LlvmIrTranslatorArm64_impl::loadOp(cs_arm64_op&, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&, llvm::Type*, bool) /home/local/devel/github/retdec/build/../src/capstone2llvmir/arm64/arm64.cpp:762:0
 #7 0x000055840416ae8d retdec::capstone2llvmir::Capstone2LlvmIrTranslator_impl<cs_arm64, cs_arm64_op>::translatePseudoAsmGeneric(cs_insn*, cs_arm64*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) /home/local/devel/github/retdec/build/../src/capstone2llvmir/capstone2llvmir_impl.cpp:2155:0
 #8 0x00005584040f81db retdec::capstone2llvmir::Capstone2LlvmIrTranslatorArm64_impl::generatePseudoInstruction(cs_insn*, cs_arm64*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) /home/local/devel/github/retdec/build/../src/capstone2llvmir/arm64/arm64.cpp:1130:0
 #9 0x00005584040f4e98 retdec::capstone2llvmir::Capstone2LlvmIrTranslatorArm64_impl::translateInstruction(cs_insn*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) /home/local/devel/github/retdec/build/../src/capstone2llvmir/arm64/arm64.cpp:344:0
#10 0x000055840416170f retdec::capstone2llvmir::Capstone2LlvmIrTranslator_impl<cs_arm64, cs_arm64_op>::translateOne(unsigned char const*&, unsigned long&, retdec::common::Address&, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) /home/local/devel/github/retdec/build/../src/capstone2llvmir/capstone2llvmir_impl.cpp:262:0
#11 0x0000558403a99e1e retdec::bin2llvmir::Decoder::translate(std::pair<unsigned char const*, unsigned long>&, retdec::common::Address&, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:349:0
#12 0x0000558403a99a88 retdec::bin2llvmir::Decoder::decodeJumpTarget(retdec::bin2llvmir::JumpTarget const&) /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:300:0
#13 0x0000558403a9914c retdec::bin2llvmir::Decoder::decode() /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:149:0
#14 0x0000558403a990a1 retdec::bin2llvmir::Decoder::run() /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:122:0
#15 0x0000558403a98f32 retdec::bin2llvmir::Decoder::runCatcher() /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:96:0
#16 0x0000558403a98e5d retdec::bin2llvmir::Decoder::runOnModule(llvm::Module&) /home/local/devel/github/retdec/build/../src/bin2llvmir/optimizations/decoder/decoder.cpp:59:0
#17 0x0000558404faab08 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/local/devel/retdec/bin/retdec-decompiler+0x1ba2b08)
#18 0x00005584039f3712 retdec::decompile(retdec::config::Config&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*) /home/local/devel/github/retdec/build/../src/retdec/retdec.cpp:518:0
#19 0x00005584039237f1 decompile(retdec::config::Config&, ProgramOptions&) /home/local/devel/github/retdec/build/../src/retdec-decompiler/retdec-decompiler.cpp:961:0
#20 0x00005584039241f9 main /home/local/devel/github/retdec/build/../src/retdec-decompiler/retdec-decompiler.cpp:1059:0
#21 0x00007f5cf9991bf7 __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:344:0
#22 0x000055840391c2ba _start (/home/local/devel/retdec/bin/retdec-decompiler+0x5142ba)
Stack dump:
0.      Program arguments: /home/local/devel/retdec/bin/retdec-decompiler at_test.elf 
1.      Running pass 'Input binary to LLVM IR decoding' on module ''.
Segmentation fault

# BOOM! Segfault
````

Here is the patch which fixes the problem, but i am sure thats not the right way of doing it, though it works :)
```shell
modified   src/capstone2llvmir/arm64/arm64.cpp
@@ -749,6 +749,8 @@ llvm::Value* Capstone2LlvmIrTranslatorArm64_impl::loadOp(
 		case ARM64_OP_REG:
 		{
 			auto* val = loadRegister(op.reg, irb);
+                        if (val == nullptr)
+                            return llvm::UndefValue::get(ty ? ty : getDefaultType());
 			auto* vec = extractVectorValue(irb, op, val);
 			auto* ext = generateOperandExtension(irb, op.ext, vec, ty);
 			return generateOperandShift(irb, op, ext);

```

**Comment by PeterMatula** [comment]
Thanks for the detailed description.

You are right, it should not be a problem to fix it - I will take a look.

**Comment by PeterMatula** [comment]
I looked at it and hopefully fixed it in #1033.

At first, I tried to understand why this happens and if we can solve it better - i.e. not just supplying `llvm::UndefValue` if we cannot generate the register. But this was not successful. According to [this](https://developer.arm.com/documentation/ddi0595/2021-06/AArch64-Instructions/AT-S1E1R--Address-Translate-Stage-1-EL1-Read), `AT S1E1R` is some weird instruction I don't even fully understand. But the main problem is that the Capstone version we use says there are 2 operands, the first being of type `ARM64_OP_SYS` with the value `ARM64_SYSREG_INVALID`. So either Capstone does not handle it correctly, or RetDec cannot interpret it as it should. I tried the `next` branch of Capstone, and it says there is only one operand. I'm however not sure about the status of their `next` branch and if they ever plan to deploy it. In any case, I don't think we can handle it much better than you suggested at this time.

I basically used your solution but put it in several more places. I tried to put it directly to `loadRegister()` and never return `nullptr` from there, but some code relies on it returning `nullptr`, so this is probably the best solution without too much refactoring.

**Comment by PeterMatula** [comment]
Fixed in #1033
